routingv8/request.go

```diff
@@ -0,0 +1,327 @@
+package routingv8
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+type CalculateMatrixBody struct {
+	// Origins defining start points of the routes in the matrix.
+	// See https://developer.here.com/documentation/matrix-routing-api/8.6.0/dev_guide/topics/modes/modes.html
+	// for guidance on the matrix limitations.
+	Origins []*GeoWaypoint `json:"origins"`
+	// Destinations defining destinations of the routes in the matrix.
+	// See https://developer.here.com/documentation/matrix-routing-api/8.6.0/dev_guide/topics/modes/modes.html
+	// for guidance on the matrix limitations.
+	Destinations []*GeoWaypoint `json:"destinations"`
+	// Time of departure for all origins. Default to now.
+	DepartureTime string `json:"departureTime,omitempty"`
+	// Region of where the matrix should be calculated.
+	RegionDefinition RegionDefinition `json:"regionDefinition"`
+	// Profile to use for route calculation in the matrix.
+	Profile Profile `json:"profile,omitempty"`
+	// Route optimization.
+	RoutingMode RoutingMode `json:"routingMode,omitempty"`
+	// Vehicle to use.
+	TransportMode TransportMode `json:"transportMode,omitempty"`
+	// Attributes to receive back in the response.
+	MatrixAttributes *MatrixAttributes `json:"matrixAttributes,omitempty"`
+	// Truck configuration
+	Truck *Truck `json:"truck,omitempty"`
+}
+
+type CalculateMatrixRequest struct {
+	// Asynchronous flag requires the client to poll the calculation results and finally requesting to download
+	// the calculation results.
+	Async Async
+	// Body to pass to request to Here Maps API
+	Body *CalculateMatrixBody
+}
+
+type GeoWaypoint struct {
+	Lat  float64 `json:"lat,omitempty"`
+	Long float64 `json:"lng,omitempty"`
+}
+
+type Profile int
+
+const (
+	ProfileUnspecified = iota
+	// Car with fast routing mode.
+	ProfileCarFast
+	// Car with short routing mode.
+	ProfileCarShort
+	// Truck with fast routing mode.
+	ProfileTruckFast
+	// Pedestrian transport mode.
+	ProfilePedestrian
+	// Bicycle transport mode.
+	ProfileBicycle
+)
+
+func (p *Profile) String() string {
+	return [...]string{"unspecified", "carFast", "carShort", "truckFast", "pedestrian", "bicycle", "invalidValue"}[*p]
+}
+
+func (p *Profile) Parse(value string) error {
+	switch value {
+	case "carFast":
+		*p = ProfileCarFast
+	case "carShort":
+		*p = ProfileCarShort
+	case "truckFast":
+		*p = ProfileTruckFast
+	case "pedestrian":
+		*p = ProfilePedestrian
+	case "bicycle":
+		*p = ProfileBicycle
+	default:
+		return fmt.Errorf("invalid profile")
+	}
+	return nil
+}
+
+func (p *Profile) MarshalJSON() ([]byte, error) {
+	buffer := bytes.NewBufferString(`"`)
```

Comment:
If we're using a buffer here to make the marshaling more efficient, we should pre-allocate the right amount of capacity to make sure that we don't trigger multiple allocations

Reply:
Sounds reasonable! How would I best go about and pre-allocate memory for a buffer?

Reply:
if we're not worried about allocations I would do:

~~~go
return []byte(strconv.Quote(p.String()))
~~~

otherwise

```go
s := p.String()
var buffer bytes.Buffer
buffer.Grow(len(s)+2)
buffer.WriteByte(`"`)
buffer.WriteString(s)
buffer.WriteByte(`"`)
return buffer.Bytes()
```